### ISSUE TITLE
Primary Window Layout

### ISSUE THREAD

**Issue by jannuary** [glimpse-editor/glimpse-nx-design#4]
### Introduction 
It is crucial to agree on the basic layout and navigation in the early stage to which more elements can be added.
Here, I am proposing a design with two sidebars (think of applications like GNU IMP, Figma, Akira, etc.) that act as a blank slate with dialogs (user-set or depending on the current persona) to be in.
### Goals
- Be intuitive and quick to understand even for the first time users
- Be modular and customizable to accommodate different layouts and workflows
- Be usable on touch-only devices
### Tentative Design
![window](https://user-images.githubusercontent.com/27908024/96106883-eb305b00-0f05-11eb-9d9d-74c3cd7cc305.png)

**Comment by ror6ax** [comment]
Sidebars on both sides are something I'm not a fan of. Unless a user chose to move something, it implies there's a reason for separating things that go on the left or on the fight - but there's no logical reason for such separation IMO.

**Comment by cauebs** [comment]
@ror6ax Well, if you just pile them up on one side it will feel cluttered and will be more difficult to find things. There might be no obvious reason to put something on one side or the other, but I think it's possible to find a criteria, e.g.:
- things that are happening, things you are _doing_, on the left. Tools, their parameters, stuff like that;
- things about how the document is organized, on the right. Layers, history, color channels, etc.

**Comment by LeoWattenberg** [comment]
While PS, GNU IMP, paint.NET et al. do have a 2-sidebar-design, it always throws me on a search for functions. 

For example, if I want to turn green to transparency – 
* is it a tool, maybe together with the eraser? 
* is it in one of the many undescript buttons in the layer thing on the right-hand side?
* is it in the right-click menu when right-clicking a layer? 
* Is it in the top menu, and if so, is it a filter, or is it under color, or is it a layer thing? 

(answer for GIMP is top menu, layer → transparency. Or color → color to alpha, and after you perform the operation, it shows up in the recently used filters. For photoshop, it's whatever [this is](https://community.adobe.com/t5/photoshop/make-one-color-of-an-image-transparent/td-p/9263056?page=1), apparently)

All of them do follow @cauebs' proposed division with tools (things you select and then click on the canvas on to apply) to one side, and document structure to the other, but there simply are many more categories which probably should exist in a semi-professional image editor, such as color info (histogram, vectorscopes, etc), generators (think GIMP's filter → render options, or blender's procedual texture nodes), various filters which apply to the entire layer (everything in GIMP's color and filter sections), and so on. All the current solutions just bunch these additional things into the top menu, sometimes indirectly into the "Window" section, so you can't even find the option you're looking for unless you know in which window/dialog it'd live. 

Assuming GlimpseNX will have a comparable feature richness to PS and GIMP eventually, I _severely_ doubt there's going to be a sensible way to categorize everything as belonging to one vs the other side. 

Because of this, I'd like to propose two alternatives:

* A workspace-oriented layout, like blender or Adobe Premiere have. This allows to have separate views for drawing/photo editing/animation/..., which all are sensible by default, but which the user can freely adapt with whatever additional windows/dialogs they need.
* A ribbon-like layout, like Microsoft Word has. This allows you to keep everything in one mode, but it allows for more explanation and sub-categorization for the individual features than the drop-downs you know from GIMP and Photoshop do. 

I'd also like to point out Microsoft Paint 3D as a source of inspiration, [which was designed as a touch-mouse hybrid, and easy enough as to be used by kids](https://www.martinkeary.com#/microsoft-paint-3d/) – so that's an image editor that at least meets 2 of your 3 goals.

**Comment by jannuary** [comment]
> A workspace-oriented layout, like blender or Adobe Premiere have. This allows to have separate views for drawing/photo editing/animation/..., which all are sensible by default, but which the user can freely adapt with whatever additional windows/dialogs they need.

@LeoWattenberg A workspace-oriented layout is planned: there will be multiple "personas", each catering to a specific workflow (e.g. "Retouch" persona could have filters and adjustment, "Color Grade" - options for color adjustment, such as a histogram) 
![Screenshot from 2020-10-17 09-20-08](https://user-images.githubusercontent.com/27908024/96329974-eab9d080-107b-11eb-9ba9-3e94d16c2ae0.png)

> While PS, GNU IMP, paint.NET et al. do have a 2-sidebar-design, it always throws me on a search for functions.

> Sidebars on both sides are something I'm not a fan of. Unless a user chose to move something, it implies there's a reason for separating things that go on the left or on the fight - but there's no logical reason for such separation IMO.

@ror6ax I don't think it would be particularly elegant nor useful to reduce the navigation to a single sidebar, but having two UI elements do basically the same thing on both sides of the window would be rather confusing.
I propose redesigning the left sidebar to not be a blank slate for whatever dialogs, but instead a non-customizable dynamic sidebar that houses options for the current tool (be it brush size, font size, outline, shadows, etc.), transform controls, as well as possibly tool presets. Here's a quick illustration to demonstrate: 
![SideBar](https://user-images.githubusercontent.com/27908024/96330210-c5c65d00-107d-11eb-90f1-ba60b447df83.png)

**Comment by ror6ax** [comment]
Personas are a very interesting idea, I'm interested to see how that turns out.

re tollbar - what if the currently selected tool does not have controls? PS has this "contextual toollbar" stretched in a horizontal bar on top of the screen. That leaves more screen space for the actual image.

**Comment by taciturasa** [comment]
Chiming in with some of my own opines and decisions with my own personal mockups.

- There is a LOT of content with a photo editor, which is why pretty much every editor uses a dual pane design. I will admit, however, that this is confusing.\
\
My own decision making for what goes on each side was that tools (which would be used to manipulate) would go on the left side, while collections or "panels" (which would collect "objects", for lack of a better logical basepoint) go on the right.\
\
This means that selecting, drawing, morphing, etc. would be on the right, while things such as palletes, layers, edit history, etc. are on the right.

The reason specifically I wanted tools and their options to be on the left is this:

- Although having a single line tool pallete saves space, it can be particularly hard to parse and navigate, especially on smaller screen sizes where it gets pushed into an overflow menu. Although I don't have mockups, my concept was that you could have an expanded view with multiple tool columns, with the tool options underneath, or a single line view, where you can click after selction to access an options popup that can also be pinned to a floating view.\
\
This is why, in my opinion, it would be a bad idea to mix panels and tools. This design would **also let you pin to the tool palette commonly used filters or manipulation tools**, since they count as an action to manipulate an image.\
\
Likely, I'd want the collapsed palette to look similar to this design, while the expand would likely look similar to the current GNU I.M.P. design (just without the panel tabs for tool options, since it would no longer be a panel).

This is all stuff that I may be repeating, since I am currently busy and cannot be fully caught up. Hopefully some of this can be of help though

**Comment by jannuary** [comment]
Adjusted some more things in my design: 
![primary-window-jannuary](https://user-images.githubusercontent.com/27908024/97573592-966c0480-1a1c-11eb-8ae3-e3d5f74a8645.png)

>     * Although having a single line tool pallete saves space, it can be particularly hard to parse and navigate, especially on smaller screen sizes where it gets pushed into an overflow menu. Although I don't have mockups, my concept was that you could have an expanded view with multiple tool columns, with the tool options underneath, or a single line view, where you can click after selction to access an options popup that can also be pinned to a floating view.

@polyjitter I feel a multi-column tool grid would be neither visually appealing nor particularly easier/more intuitive. 
My design could provide a handy mental model - one could think of the tool options panel in terms of a tabbed/ribbon interface.

Pinning filters and such to the tool palette is something to be considered, though that also rises a question on how will this be realized.

>  PS has this "contextual toollbar" stretched in a horizontal bar on top of the screen.

@ror6ax While it's true PS' approach saves horizontal space, I feel that tiny sliver looks cramped and isn't particularly useful. Here's a screenshot of the "Text" tool options in Inkscape, which adopts a similar design to what PS has:
![Screenshot from 2020-10-29 19-38-04](https://user-images.githubusercontent.com/27908024/97574963-940aaa00-1a1e-11eb-8c38-36449bec211d.png)

**Comment by malabo1990** [comment]
text size I would like to change to slider so you can modify it as fast as the pencil properties slider.
